### PR TITLE
Feature/dockstore cwls

### PR DESCRIPTION
**Pull request by byb121**
@lbarbon cgpRna is loaded with command line tools, but in order to complete a rna-seq analysis, many commands need to run. I grouped these commands into several sets and added a python command line tool `run-cgprna` as a CLI to use them. Upon this CLI, CWL files can be built on as well. 

So far, through the CLI, we can do: map, generate QC stats, bigwig coverage file and count reads mapped to genes. I intend to write a CWL for each of the function, and probably uses a CWL workflow to chain them later.

Please have a look at the python code (ignore Dockerfile and CWLs).

I have several things I'm not too sure about:
1. does the code look ok? I tested them, worked fine, but probably missed edge cases.
1. if I want to write some tests for the python code, what should I test against, as its function is a bit simple.

Any thought on anything from you is welcome :)

### CODE REVIEW

File: run-cgprna/run_cgprna/map.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            BAM_INDEX_TEMPLATE
+        ],
+        params)
+    
+    # clean temp dir
+    if clean_temp:
+        shutil.rmtree(temp_dir)
+
+
+def find(pattern, path):
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/map.py]: This could be rewritten as a chaining of generators.

File: run-cgprna/run_cgprna/map.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                os.symlink(
+                    gtfs[0],
+                    os.path.join(final_gtf_folder, ref_related_args['--gene-build-gtf-name'])
+                )
+            else:
+                sys.exit('Error: none or too many GTF files in refence bundle. Found GTF(s): %s' % ','.join(gtfs))
+        else:
+            sys.exit('Error: wrong input format. "--reference" can only be a tar.gz file or a folder.')
+    else:
+        if any([ele is None for ele in ref_related_args.values()]):
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/map.py]: The square brackets are not necessary in this context; dropping them would also make this evaluation lazy.

File: run-cgprna/run_cgprna/map.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    reference_data_root = os.path.abspath(args.ref)
+    # Anything not a file will be treated as a reference root folder
+    if os.path.isfile(reference_data_root):
+        # If a pre-built ref bundle tar file is supplied
+        if re.match(r'.*\.tar\.gz$', os.path.basename(reference_data_root)):
+            mkdir(temp_dir)
+            clean_temp = 1
+
+            # use the tar file name as the ref_root_dir_name, so that in BAM header, people can tell which bundle file was used
+            ref_root_dir_name = re.match(r'(.*)\.tar\.gz$', os.path.basename(reference_data_root)).group(1)
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/map.py]: Would not this capture the file name as well? Maybe just os.dirname?

File: run-cgprna/run_cgprna/map.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            # find the GTF file
+            gtfs = find('*.gtf', bundle_decompress_path)
+            if len(gtfs) == 1:
+                ref_related_args['--gene-build-gtf-name'] = os.path.basename(gtfs[0])
+                # link the file to final_gtf_folder
+                os.symlink(
+                    gtfs[0],
+                    os.path.join(final_gtf_folder, ref_related_args['--gene-build-gtf-name'])
+                )
+            else:
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/map.py]: I would suggest [guard clauses](https://www.refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) instead of `else` statements when returning or exiting, to improve readability.

File: run-cgprna/run_cgprna/map.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    # prepare the output dir
+    mkdir(args.out_dir)
+
+    # prepare the reference
+    reference_data_root = os.path.abspath(args.ref)
+    # Anything not a file will be treated as a reference root folder
+    if os.path.isfile(reference_data_root):
+        # If a pre-built ref bundle tar file is supplied
+        if re.match(r'.*\.tar\.gz$', os.path.basename(reference_data_root)):
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/map.py]: Or `reference_data_root .endswith('.tar.gz')`.

File: run-cgprna/run_cgprna/map.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        other_options.append('-library "%s"' % args.lb_tag)
+    if args.ds_tag:
+        other_options.append('-ds-tag "%s"' % args.ds_tag)
+    if args.pl_tag:
+        other_options.append('-machine-type "%s"' % args.pl_tag)
+    if args.pu_tag:
+        other_options.append('-npg-run "%s"' % args.pu_tag)
+
+    # prepare the output dir
+    mkdir(args.out_dir)
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/map.py]: It's going to raise an exception if the directory already exists. This may be a reasonable behaviour, but if you want to be more lenient: `os.makedirs(fp, exist_ok=True)`.

File: run-cgprna/run_cgprna/mapping_stats.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    sample_name, _ = os.path.splitext(os.path.basename(args.input))
+
+    # prepare the output dir
+    mkdir(args.out_dir)
+
+    # prepare the reference
+    reference_data_root=os.path.abspath(args.ref)
+    # If a ref bundle tar file is supplied
+    # Anything else will be treated as a reference folder
+    if re.search(r'\.tar\.gz$', args.ref):
```
Comment (lbarbon) [inline run-cgprna/run_cgprna/mapping_stats.py]: Or `args.ref.endswith('.tar.gz')`.

File: run-cgprna/run_cgprna/map.py (hunk 1 above)
Comment (byb121) [reply run-cgprna/run_cgprna/map.py]: Tried and love it!

File: run-cgprna/run_cgprna/map.py (hunk 3 above)
Comment (byb121) [reply run-cgprna/run_cgprna/map.py]: hmm, the variable name is a bit confusing. The line is trying to capture tar file base name without the extensions, so that the variable can be later on used to name the folder where reference files are extracted into.

File: run-cgprna/run_cgprna/map.py (hunk 6 above)
Comment (byb121) [reply run-cgprna/run_cgprna/map.py]: the line uses mkdir imported from __init__.py, very confusing, isn't it? :)
